x/connectivity/connectivity.go

```diff
@@ -78,15 +100,15 @@ func isTimeout(err error) bool {
 	return errors.As(err, &timeErr) && timeErr.Timeout()
 }
 
-func makeConnectivityError(op string, err error) *ConnectivityError {
+func makeConnectivityError(op string, startTime time.Time, duration time.Duration, err error) *ConnectivityError {
```

Comment:
This is unnecessary coupling, please revert. The connectivity error semantic is just what was observed. The time and duration should go in the report instead, or the holder of the connectivity error. The Outline client doesn't care about the start time or the duration, for instance.

Reply:
Thanks for the feedback. 

The `startTime` and `duration` params here are specific to the low-level `connect` (dial), `send` (write), `receive` (read) operations which offer higher resolution timestamps as opposed to time-stamping the overall DNS resolution operation. 

We discussed this briefly [here](https://github.com/Jigsaw-Code/outline-sdk/pull/140#discussion_r1430618543). 

Do you mean instead of including these params in connectivity error data structure, I define a parent data structure such as `ConnectivityReport` here that includes `connectivityError`, and the operation specific timestamps?

Shall we include the timestamps for all operations for the success cases as well? For example something like this:

```
type ConnectivityReport struct {
        protocol string
        connect OpTimestamp
        send OpTimestamp
        receive OpTimestamp
        connectivityError *ConnectivityError
}

type OpTimestamp struct {
	StartTime time.Time
	Duration time.Duration
}
```

Reply:
The per-operation time stamps turned out to be a bit of a pain: they require invasive changes to the dialers and connections. I think there are workarounds, but the reality is that we don't really need them for now. It's good enough

We should focus on making progress on the bigger tasks.

One thing we don't have yet is the reporting of the transport. Let's add that to the test-connectivity tool.

It would also be helpful to have the reporting in the gui app, so people can test it in the field.

Then we should start looking into adding to an actual circumvention tool. A fork of the Outline Client would make sense, but we should look into Tor as well.

Reply:
> The per-operation time stamps turned out to be a bit of a pain: they require invasive changes to the dialers and connections. I think there are workarounds, but the reality is that we don't really need them for now. What we have is good enough. We should focus on making progress on the bigger tasks.

I think there was a misunderstanding on this since I was under the impression that we were still pursuing the timestamp idea but not using spans/traces...

I agree that the amount of useful information we can get from timestamps is minimal. Maybe we can revisit this later. Just to clarify, the reason I added timestamps to `connectivityError` was to avoid making changes to the `TestConnectivityWithResolver` interfaces so that we don't have to return other params and break things elsewhere. The other way to pass the timestamp info was through `context.Context` which I did [here](https://github.com/amircybersec/outline-sdk/blob/simple-trace/x/connectivity/tracer.go) with custom span/trace.

> One thing we don't have yet is the reporting of the transport. Let's add that to the test-connectivity tool.

The proceeding task for this was to define config sanitizer. I just pushed the code for `SanitizeConfig` that I added to `config.go` package. I am going to add the logic to the `test-connectivity` to include the transport info. 

> It would also be helpful to have the reporting in the gui app, so people can test it in the field.

This has been actually added to GUI app but I need to refactor the code to use the new `SanitizeConfig` method

> Then we should start looking into adding to an actual circumvention tool. A fork of the Outline Client would make sense, but we should look into Tor as well. Also the other clients out there (v2ray, xray, etc...)

I think a good starting point for this is to use Outline CLI on Linux. For v2ray, xray core, etc we need to have support for `vless://`, `vmess://`, etc transports but since these client apps support shadowsocks, we can demo connectivity tests with ss:// only initially